Enhancement/7901 add source to ri mixpanel

Miraeld opened the pull request:
# Description

Fixes #7901

This enhancement adds source tracking to Rocket Insights Mixpanel events, enabling product analytics to understand where users initiate actions. The feature tracks user interaction context across all Rocket Insights entry points.


## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## Detailed scenario

### What was tested

Triggering every possible way to add page and perform tests:

<details>
<summary>Results here</summary>

### Adds a page from Rocket Insights tab
```
[13-Nov-2025 11:33:52 UTC] Mixpanel Tracking Data: Array
(
    [event] => Rocket Insights Page Added
    [data] => Array
        (
            [context] => wp_plugin
            [plan_type] => perf-monitor-free
            [tracked_pages] => 2
            [source] => add-on page
        )

)

[13-Nov-2025 11:34:18 UTC] Mixpanel Tracking Data: Array
(
    [event] => Rocket Insights Performance Test
    [data] => Array
        (
            [context] => wp_plugin
            [status] => completed
            [score] => 100
            [retest] => 
            [duration] => 26
            [plan_type] => perf-monitor-free
            [source] => add-on page
        )

)
```

### Add page from page listing
```
[13-Nov-2025 11:35:29 UTC] Mixpanel Tracking Data: Array
(
    [event] => Rocket Insights Page Added
    [data] => Array
        (
            [context] => wp_plugin
            [plan_type] => perf-monitor-free
            [tracked_pages] => 3
            [source] => post type listing
        )

)

[13-Nov-2025 11:36:20 UTC] Mixpanel Tracking Data: Array
(
    [event] => Rocket Insights Performance Test
    [data] => Array
        (
            [context] => wp_plugin
            [status] => failed
            [score] => 0
            [retest] => 
            [duration] => 51
            [plan_type] => perf-monitor-free
            [source] => post type listing
        )

)
```
### Click on `Add Homepage` from WP Rocket dashboard
```
[13-Nov-2025 11:51:38 UTC] Mixpanel Tracking Data: Array
(
    [event] => Rocket Insights Page Added
    [data] => Array
        (
            [context] => wp_plugin
            [plan_type] => perf-monitor-free
            [tracked_pages] => 1
            [source] => dashboard
        )

)

[13-Nov-2025 11:52:04 UTC] Mixpanel Tracking Data: Array
(
    [event] => Rocket Insights Performance Test
    [data] => Array
        (
            [context] => wp_plugin
            [status] => failed
            [score] => 0
            [retest] => 
            [duration] => 26
            [plan_type] => perf-monitor-free
            [source] => dashboard
        )

)
```
</details>

### How to test

Perform actions on Rocket Insights that would trigger Mixpanel tracking while having Rocket Analytics enabled.
Note that `auto-added homepage` is impossible to trigger as Rocket Analytics is disabled by default.

## Technical description

### Architecture Overview

The implementation adds source tracking across the entire Rocket Insights workflow:

**1. Frontend Source Detection**
- HTML templates include `data-source` attributes on all interaction buttons
- JavaScript detects source from button context and passes to API calls
- Sources: `dashboard`, `post type listing`, `add-on page`

**2. REST API Parameter Handling**
- New optional `source` parameter with validation against allowed values
- Default value: `add-on page` (for direct API calls)
- Backward compatibility maintained

**3. Database Persistence**
- Source stored in `data` JSON field for performance test attribution
- Survives async SaaS processing workflow
- Enables delayed event tracking with original context

**4. Mixpanel Event Enhancement**
- `Rocket Insights Page Added`: Immediate source tracking
- `Rocket Insights Performance Test`: Source retrieved from database

### Source Types

| Source | Context | Tracking |
|--------|---------|----------|
| `dashboard` | Global Score Widget buttons | ✅ Active |
| `post type listing` | Posts/Pages/CPT admin columns | ✅ Active |
| `add-on page` | Rocket Insights tab interface | ✅ Active |
| `auto-added homepage` | Fresh install auto-add | ❌ Impossible* |
| `performance monitoring` | CRON scheduled retests | ✅ Active |

*`auto-added homepage` is impossible to trigger because Rocket Analytics is disabled by default on fresh installations.



**Performance Test Events**: Any events that could be tracked have `performance monitoring` as default value, ensuring CRON retests are properly attributed to scheduled monitoring activities.

### Documentation

The implementation follows WP Rocket's architectural patterns:
- **Service Provider Pattern**: Dependency injection for all services
- **Subscriber Pattern**: WordPress hooks managed via Event_Manager
- **Database Layer**: JSON data field for metadata persistence
- **Type Safety**: Parameter validation and sanitization throughout


# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Unticked items justification
No test for this.

Miraeld committed:
feat: Add source parameter to Rocket Insights REST API

- Add source parameter validation with allowed values: dashboard, post type listing, add-on page, auto-added homepage, performance monitoring
- Make source parameter optional with default value 'add-on page'
- Update create_item method to accept and validate source parameter
- Pass source parameter through rocket_rocket_insights_job_added hook

Ref: #7901

Miraeld committed:
feat: Implement source tracking in Mixpanel events

- Update track_rocket_insights_url_added() to accept source parameter
- Update track_rocket_insights_test() to include source from database
- Add source to 'Rocket Insights Page Added' event data
- Add source to 'Rocket Insights Performance Test' event data
- Maintain backward compatibility with optional source parameter

Both Mixpanel events now include source attribution for analytics.

Ref: #7901

Miraeld committed:
feat: Update Controller and Subscriber for source tracking

Controller changes:
- Add source parameter to add_homepage() method with default 'auto-added homepage'
- Update add_homepage_from_widget() to pass 'dashboard' source
- Store source in database data field for persistent attribution
- Update hook calls to pass dynamic source parameter

Subscriber changes:
- Update hook subscription to handle 4 parameters including source
- Preserve original source in retest_all_pages() or default to 'performance monitoring'
- Maintain source attribution through scheduled retests

Fixes dashboard homepage source issue - now correctly tracks as 'dashboard'.

Ref: #7901

Miraeld committed:
feat: Add data-source attributes to HTML templates

- Add data-source='post type listing' to post/page listing buttons
- Add data-source='add-on page' to Rocket Insights tab buttons
- Add data-source='dashboard' to global score widget buttons
- Update all Test/Add Page buttons with appropriate source attribution
- Enable JavaScript source detection for API calls

Templates now include source data for frontend tracking integration.

Ref: #7901

Miraeld committed:
feat: Implement JavaScript source detection and API integration

- Update addNewPage() in rocket-insights.js to detect source from data attributes
- Update handleAddPage() in ajax.js to extract and pass source to API calls
- Add source parameter to REST API requests
- Enable dynamic source detection from button context
- Integrate with existing AJAX and form submission workflows

JavaScript now detects user interaction context and passes source to backend.

Ref: #7901

Miraeld committed:
fix: Address GitHub Copilot review feedback

- Fix JavaScript variable declaration order (column before source)
- Correct fallback values to match REST API allowed values format
- Use 'post type listing' and 'add-on page' instead of underscore variants
- Ensure consistency between JavaScript and PHP default values

Miraeld committed:
Adds missing js

Miraeld commented on inc/Engine/Admin/RocketInsights/Rest.php:
```diff
@@ -184,6 +184,21 @@ public function register_routes(): void {
+							'sanitize_callback' => function ( $param ) {
+								return sanitize_text_field( $param );
+							},
+							'default'           => 'addon_page',
```
Fixed

Miraeld commented on src/js/global/ajax.js:
```diff
@@ -514,6 +514,7 @@ document.addEventListener('DOMContentLoaded', function() {
 		}
 
 		const pageUrl = $pageUrlInput.val().trim();
+		const source = $(this).data('source') || 'addon_page';
```
Fixed

Miraeld commented on src/js/global/rocket-insights.js:
```diff
@@ -113,7 +115,10 @@ module.exports = (function () {
-			data: { page_url: url },
+			data: { 
+				page_url: url,
+				source: source || 'post_listing'
```
Fixed

Miraeld commented on inc/Engine/Admin/RocketInsights/Rest.php:
```diff
@@ -330,7 +345,8 @@ public function create_item( $request ) {
 
-		$url = $payload['processed_url'];
+		$url    = $payload['processed_url'];
+		$source = $request->get_param( 'source' ) ?: 'addon_page';
```
Fixed

Miraeld commented on src/js/global/rocket-insights.js:
```diff
@@ -39,6 +39,7 @@ module.exports = (function () {
 			e.preventDefault();
 			const button = jQuery(this);
 			const url = button.data('url');
+			const source = button.data('source') || column.data('source') || 'post_listing';
```
Fixed

Miraeld commented on assets/js/rocket-insights.js:
```diff
@@ -42,13 +42,14 @@ module.exports = function () {
       const button = jQuery(this);
       const url = button.data('url');
+      const source = button.data('source') || column.data('source') || 'post_listing';
       const column = button.closest('.wpr-ri-column');
```
Fixed

Miraeld commented on src/js/global/rocket-insights.js:
```diff
@@ -39,6 +39,7 @@ module.exports = (function () {
 			const button = jQuery(this);
 			const url = button.data('url');
+			const source = button.data('source') || column.data('source') || 'post_listing';
 			const column = button.closest('.wpr-ri-column');
```
Fixed

Miraeld commented on src/js/global/rocket-insights.js:
```diff
@@ -39,6 +39,7 @@ module.exports = (function () {
 			const button = jQuery(this);
 			const url = button.data('url');
+			const source = button.data('source') || column.data('source') || 'post_listing';
 			const column = button.closest('.wpr-ri-column');
```
Fixed

Miraeld commented on assets/js/rocket-insights.js:
```diff
@@ -42,13 +42,14 @@ module.exports = function () {
       const button = jQuery(this);
       const url = button.data('url');
+      const source = button.data('source') || column.data('source') || 'post_listing';
       const column = button.closest('.wpr-ri-column');
```
Fixed

Miraeld committed:
Fix phpcs

Miraeld committed:
fix tests

jeawhanlee commented on inc/Engine/Admin/RocketInsights/Rest.php:
```diff
@@ -330,7 +345,8 @@ public function create_item( $request ) {
 
-		$url = $payload['processed_url'];
+		$url    = $payload['processed_url'];
+		$source = $request->get_param( 'source' ) ?: 'add-on page';
```
Having fallback values like this for variables that should always have a value especially in the context of APIs can be misleading for debugging in the future supposing an empty value is passed in js or something like this we would always get add-on page which might not be the actual source. I strongly recommend having this without a fallback.

Miraeld replied:
Fixed in [7f969](https://github.com/wp-media/wp-rocket/commit/7f9698370) - Removed fallback entirely and made source parameter required. Changed from `$source = $request->get_param( 'source' ) ?: 'add-on page';` to `$source = $request->get_param( 'source' );`

jeawhanlee commented on inc/Engine/Tracking/Tracking.php:
```diff
@@ -229,21 +229,28 @@ public function track_optin_change( $status ): void {
 	 * @return void
 	 */
-	public function track_rocket_insights_url_added( $url, $plan, $urls_count ): void {
+	public function track_rocket_insights_url_added( $url, $plan, $urls_count, $source = '' ): void {
```
Is there a need to make the source parameter optional, it would always be present for the hook.

Miraeld replied:
Fixed in [bfb8d](https://github.com/wp-media/wp-rocket/commit/bfb8d911c) - Made source parameter required without default values. Removed `= ''` from function signature.

jeawhanlee commented on inc/Engine/Tracking/Tracking.php:
```diff
@@ -229,21 +229,28 @@ public function track_optin_change( $status ): void {
+			'tracked_pages' => $urls_count,
+		];
+
+		if ( ! empty( $source ) ) {
```
I don't think this conditional check is also useful.

Miraeld replied:
Fixed in [bfb8d](https://github.com/wp-media/wp-rocket/commit/bfb8d911c) - Removed conditional check `if ( ! empty( $source ) )` and directly include source in Mixpanel data.

jeawhanlee commented on inc/Engine/Tracking/Tracking.php:
```diff
@@ -267,16 +274,22 @@ public function track_rocket_insights_test( $row_details, $job_details, $plan ):
+			'plan_type' => $plan,
+		];
+
+		if ( ! empty( $row_details->data['source'] ) ) {
```
I believe this check as well, will not be useful as we add the source for every scenario.

Miraeld replied:
Fixed in [bfb8d](https://github.com/wp-media/wp-rocket/commit/bfb8d911c) - Removed conditional check and directly include source in tracking data array.

jeawhanlee commented on src/js/global/rocket-insights.js:
```diff
@@ -113,7 +115,10 @@ module.exports = (function () {
-			data: { page_url: url },
+			data: { 
+				page_url: url,
+				source: source || 'post type listing'
```
We already do this check in line 43
```suggestion
				source: source
```

Miraeld replied:
Fixed in [20e60](https://github.com/wp-media/wp-rocket/commit/20e603374) - Removed redundant `|| 'post type listing'` fallback in API call since source is already determined.

jeawhanlee reviewed (changes requested):
Great Job, Sir 🙌

Miraeld committed:
Fix: Make source parameter required in REST API

- Make source parameter required instead of optional
- Remove unnecessary default fallback since source should always be provided
- Simplify validation callback by removing empty check

Miraeld committed:
Fix: Remove unnecessary fallbacks and conditions in tracking

- Remove default empty string parameter since source is always provided
- Remove conditional checks for source in Mixpanel data
- Simplify tracking functions by directly including source in data arrays

Miraeld committed:
Fix: Remove unnecessary fallback in JavaScript source parameter

- Remove redundant fallback in API call since source is already determined
- Source is guaranteed to be present from button/column data attributes

Miraeld committed:
Build: Rebuild compiled JavaScript assets

- Update compiled rocket-insights.js with source parameter fixes
- Update minified version and source maps

wordpressfan commented:
Check failed test before sending to review again.
Thanks.

Miraeld committed:
Fix tests

wordpressfan committed:
Merge branch 'develop' into enhancement/7901-add-source-to-ri-mixpanel

hanna-meda commented:
@Miraeld thank you for this PR.

Exploratory findings:
- [x] This warning logs each time we retest a URL and the source value is NULL:
>[17-Nov-2025 16:00:48 UTC] PHP Warning:  Undefined array key "source" in >/var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Tracking/Tracking.php on line 281
>[17-Nov-2025 16:00:48 UTC] Mixpanel event: Rocket Insights Performance Test array (
  >'context' => 'wp_plugin',
  >'status' => 'failed',
  >'score' => 0,
  >'retest' => true,
  >'duration' => 84,
  >'plan_type' => 'perf-monitor-free',
  >'source' => NULL,
  >'domain' => '15792f4556471b6e92774b099ef5e94e64e0184993567cdf75fd8797',
  >'wp_version' => '6.7',
  >'php_version' => '8.2',
  >'plugin' => 'wp rocket 3.20.1.2',
  >'brand' => 'wp media',
  >'application' => 'wp rocket',
>)

[UPDATE] We will go with `re-test add-on page` & `re-test post type listing` in this case. Related [slack thread](https://group-onecom.slack.com/archives/C08F4LZB2QG/p1763453924932989) (approved by PO).

- [x] For premium users, when enabling Performance Monitoring schedule and running `rocket_insights_retest` hook, it fails with the following log message: 
> action failed via Admin List Table: json_decode(): Argument #1 ($json) must be of type string, array given
Note: This is working correctly on the `develop` branch.
<img width="1335" height="366" alt="Screenshot 2025-11-19 at 02 33 19" src="https://github.com/user-attachments/assets/d1deca6e-56ec-4176-b923-f6127539f507" />

Using [this workaround snippet](https://gist.github.com/hanna-meda/3941a7cdfab7634a693d2e809c128a27) so that `rocket_insights_retest` hook runs successfully, I can confirm that the event source is correctly shown as 'performance monitoring'. 

- [x] Adding new URLs to the RI tab doesn't work. This is a regression in the PR (it works fine on the `develop` branch). The following errors appear in the Console:
<img width="1531" height="949" alt="Screenshot 2025-11-19 at 02 22 46" src="https://github.com/user-attachments/assets/3478df35-1009-41b2-927c-291c97a7eb49" />

❌ [NEW]  Source for some URLs is 're-test post type listing' when running `rocket_insights_retest` hook, or vice-versa, ‘performance monitoring’ when re-testing from Posts list.  
Steps to reproduce:
(Premium user with Performance monitoring option enabled)
1. Add one more URL on the RI tab (besides the homepage, automatically added at fresh install/activation).
2. Add another URL from the Posts list by clicking "Test the page".
3. Go to AC > Scheduled Actions and run the "rocket_insights_retest" hook. 
4. Check the Mixpanel events.  ==> Performance Test events for the homepage and the URL added from the RI tab show `performance monitoring` as the source, but the one added from the Posts list shows `re-test post type listing`. Expected: All URLs triggered by the scheduled task should show `performance monitoring` as the source.  5. Go to the Posts list and test a new URL or re-test an existing one.
1. Check the Mixpanel events again.  ==> The source may show `performance monitoring` when it should be `post type listing` or `re-test post type listing`. 
Note: This occurs approximately every other attempt when using this PR.
<img width="752" height="828" alt="Screenshot 2025-11-19 at 11 46 06" src="https://github.com/user-attachments/assets/9db96d6a-2943-4471-8f53-d76291e64fd1" />

- [x] [NEW] Sometimes when running the `rocket_insights_retest hook`, an extra Performance Test event is immediately sent to Mixpanel with status completed and score 100. So by the end of processing all URLs, there will be an extra event logged, this first one.  
Note: This occurred twice in ~8 attempts using this PR, but did not occur during testing on the develop branch, suggesting it may also be related to this PR.

Miraeld committed:
Add source tracking to Rocket Insights Mixpanel events

Miraeld commented:
Hello hello, 

So after discussing with product we decided to add two new sources: `re-test add-on page` & `re-test post type listing`. 

So now when we click on the `re-test` from Rocket Insights tab we get something similar to:
```
[18-Nov-2025 10:29:14 UTC] Mixpanel event: Rocket Insights Performance Test array (
  'context' => 'wp_plugin',
  'status' => 'failed',
  'score' => 0,
  'retest' => true,
  'duration' => 28,
  'plan_type' => 'perf-monitor-free',
  'source' => 're-test add-on page',
  'domain' => '59aa441d01b3c57fa1d6c05809cd3aebe8032dc619119b78daee3c1b',
  'wp_version' => '6.8',
  'php_version' => '8.2',
  'plugin' => 'wp rocket 3.20.1.2',
  'brand' => 'wp media',
  'application' => 'wp rocket',
)
```

And from post listing:
```
[18-Nov-2025 10:33:14 UTC] Mixpanel event: Rocket Insights Performance Test array (
  'context' => 'wp_plugin',
  'status' => 'failed',
  'score' => 0,
  'retest' => true,
  'duration' => 12,
  'plan_type' => 'perf-monitor-free',
  'source' => 're-test post type listing',
  'domain' => '59aa441d01b3c57fa1d6c05809cd3aebe8032dc619119b78daee3c1b',
  'wp_version' => '6.8',
  'php_version' => '8.2',
  'plugin' => 'wp rocket 3.20.1.2',
  'brand' => 'wp media',
  'application' => 'wp rocket',
)
```

Miraeld committed:
Fix test

jeawhanlee commented on inc/Engine/Admin/RocketInsights/Rest.php:
```diff
@@ -351,9 +360,6 @@ public function create_item( $request ) {
 			'title' => $page_title,
-			'data'  => [
-				'source' => $source,
-			],
```
Since this line has been removed, this means we no longer save the source in the DB and when a page test is completed we don't have context of the source. Just confirming if this is expected?

Miraeld replied:
I did change the file, so it lost your reference.

hanna-meda commented:
@Miraeld, I retested with the latest changes and added[ two new findings](https://github.com/wp-media/wp-rocket/pull/7903#issuecomment-3542658267) above. Please take another look. Thanks.

Miraeld committed:
Fix feedback

Miraeld commented:
@hanna-meda I've made some changes, and re-tested, seems to work now

hanna-meda commented:
@Miraeld, I re-tested and all previous points are fixed; however, two new issues were discovered, added in the same [comment](https://github.com/wp-media/wp-rocket/pull/7903#issuecomment-3542658267). Thanks.

Miraeld committed:
Fix remove original source from cron

Miraeld commented:
@hanna-meda it should be fixed from my test :) 
We were re-using the original source saved..

hanna-meda commented:
@Miraeld, after checking with the latest change, it still doesn’t fix the point marked with ❌ in my previous comment.
The new steps to reproduce are:

From RI tab:
1. Add a new URL to the RI tab ==> Page Added event at Mixpanel with source: add-on page :white_check_mark:
2. Wait until URL completed ==> Performance Test event at Mixpanel with source: performance monitoring :x: should be: add-on page
3. Re-test URL from the RI tab ==> Performance Test event at Mixpanel with source re-test add-on page :white_check_mark:

From Posts/Pages/Products/Custom posts list:
1. Add URL from Posts list ==> Page Added event at Mixpanel with source: post type listing :white_check_mark:
2. Wait until URL completed ==> Performance Test event at Mixpanel with source: performance monitoring :x: should be: post type listing
3. Re-test URL from Posts list ==> Performance Test event at Mixpanel with source re-test post type listing :white_check_mark:

P.S. Running the `rocket_insights_retest` hook, all Performance Test events will have for source: `performance monitoring` ✅ 

cc @engahmeds3ed, in case Gael will be on the add-on team for the rest of the sprint.

wordpressfan committed:
add source when adding the url

wordpressfan committed:
fix the source location

wordpressfan committed:
fix tests

wordpressfan committed:
Merge branch 'develop' into enhancement/7901-add-source-to-ri-mixpanel

wordpressfan commented:
@hanna-meda all good now, can u plz double-check?

hanna-meda commented:
Looking good now, @wordpressfan. I’m finishing the TP execution here, and if everything goes well through the end, this will be ready to merge. I’ll announce it ASAP.

hanna-meda reviewed (approved):
QA done. Everything is looking good. Thanks, @wordpressfan, @Miraeld.
[testrail-report-790.pdf](https://github.com/user-attachments/files/23652317/testrail-report-790.pdf)
